// @vuepress/shared-utils/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.semver = exports.escapeHtml = exports.hash = exports.globby = exports.path = exports.fs = exports.chalk = exports.unescapeHtml = exports.tryChain = exports.toAbsolutePath = exports.sort = exports.slugify = exports.performance = exports.parseVueFrontmatter = exports.parseHeaders = exports.parseFrontmatter = exports.parseEmojis = exports.normalizeConfig = exports.moduleResolver = exports.moduleLoader = exports.logger = exports.isIndexFile = exports.inferTitle = exports.getPermalink = exports.fileToPath = exports.fallback = exports.extractHeaders = exports.env = exports.ensureLeadingSlash = exports.ensureEndingSlash = exports.deeplyParseHeaders = exports.datatypes = exports.compose = exports.codegen = void 0;
const codegen = __importStar(require("./codegen"));
exports.codegen = codegen;
const compose_1 = __importDefault(require("./compose"));
exports.compose = compose_1.default;
const datatypes = __importStar(require("./datatypes"));
exports.datatypes = datatypes;
const deeplyParseHeaders_1 = __importDefault(require("./deeplyParseHeaders"));
exports.deeplyParseHeaders = deeplyParseHeaders_1.default;
const ensureEndingSlash_1 = __importDefault(require("./ensureEndingSlash"));
exports.ensureEndingSlash = ensureEndingSlash_1.default;
const ensureLeadingSlash_1 = __importDefault(require("./ensureLeadingSlash"));
exports.ensureLeadingSlash = ensureLeadingSlash_1.default;
const env_1 = __importDefault(require("./env"));
exports.env = env_1.default;
const extractHeaders_1 = __importDefault(require("./extractHeaders"));
exports.extractHeaders = extractHeaders_1.default;
const fallback = __importStar(require("./fallback"));
exports.fallback = fallback;
const fileToPath_1 = __importDefault(require("./fileToPath"));
exports.fileToPath = fileToPath_1.default;
const getPermalink_1 = __importDefault(require("./getPermalink"));
exports.getPermalink = getPermalink_1.default;
const inferTitle_1 = __importDefault(require("./inferTitle"));
exports.inferTitle = inferTitle_1.default;
const isIndexFile = __importStar(require("./isIndexFile"));
exports.isIndexFile = isIndexFile;
const logger_1 = __importDefault(require("./logger"));
exports.logger = logger_1.default;
const moduleLoader = __importStar(require("./moduleLoader"));
exports.moduleLoader = moduleLoader;
const moduleResolver = __importStar(require("./moduleResolver"));
exports.moduleResolver = moduleResolver;
const normalizeConfig_1 = __importDefault(require("./normalizeConfig"));
exports.normalizeConfig = normalizeConfig_1.default;
const parseEmojis = __importStar(require("./parseEmojis"));
exports.parseEmojis = parseEmojis;
const parseFrontmatter_1 = __importDefault(require("./parseFrontmatter"));
exports.parseFrontmatter = parseFrontmatter_1.default;
const parseHeaders_1 = __importDefault(require("./parseHeaders"));
exports.parseHeaders = parseHeaders_1.default;
const parseVueFrontmatter = __importStar(require("./parseVueFrontmatter"));
exports.parseVueFrontmatter = parseVueFrontmatter;
const performance_1 = __importDefault(require("./performance"));
exports.performance = performance_1.default;
const slugify_1 = __importDefault(require("./slugify"));
exports.slugify = slugify_1.default;
const sort_1 = __importDefault(require("./sort"));
exports.sort = sort_1.default;
const toAbsolutePath_1 = __importDefault(require("./toAbsolutePath"));
exports.toAbsolutePath = toAbsolutePath_1.default;
const tryChain_1 = __importDefault(require("./tryChain"));
exports.tryChain = tryChain_1.default;
const unescapeHtml_1 = __importDefault(require("./unescapeHtml"));
exports.unescapeHtml = unescapeHtml_1.default;
const chalk_1 = __importDefault(require("chalk"));
exports.chalk = chalk_1.default;
const fs_extra_1 = __importDefault(require("fs-extra"));
exports.fs = fs_extra_1.default;
const path_1 = __importDefault(require("path"));
exports.path = path_1.default;
const globby_1 = __importDefault(require("globby"));
exports.globby = globby_1.default;
const hash_sum_1 = __importDefault(require("hash-sum"));
exports.hash = hash_sum_1.default;
const escape_html_1 = __importDefault(require("escape-html"));
exports.escapeHtml = escape_html_1.default;
const semver_1 = __importDefault(require("semver"));
exports.semver = semver_1.default;
