// vuepress-plugin-tags/clientRootMixin.js
import Vue from 'vue'
import tags from './tags'

const defaultConfig = {
  type: 'default', // 标签预定义样式
  color: '#42b983',  // 标签字体颜色
  border: '1px solid #e2faef', // 标签边框颜色
  backgroundColor: '#f0faf5', // 标签背景颜色
  selector: '.page .content__default h1' // 你要将此标签渲染挂载到哪个元素后面？默认是第一个 H1 标签后面
}


export default {
  created () {
    this._mrcodeTags = {
      config: Object.assign({}, defaultConfig, MRCODE_TAGS)
    }
  },
  mounted () {
    this.$nextTick(() => {
      setTimeout(() => {
        this._addTagsToPage()
      }, 300)
    })
  },
  watch: {
    '$page.path' (val) {
      this.$nextTick(() => {
        this._addTagsToPage()
      })
    },
    '$page.key' (val) {
    }
  },
  updated () {

  },
  methods: {
    _addTagsToPage () {
      const h1s = document.querySelectorAll(this._mrcodeTags.config.selector)
      if (h1s.length == 0) {
        return
      }
      if (!this.$frontmatter.tags) {
        return
      }
      let config = this._mrcodeTags.config
      let h1 = h1s[0]
      let Tags = Vue.extend(tags)
      let tags1 = new Tags({
        propsData: {
          tags: this.$frontmatter.tags,
          config: config
        }
      })
      tags1.$mount()

      insterAfter(tags1.$el, h1)
    },
  }
}

function insterAfter (newElement, targetElement) {
  let parent = targetElement.parentNode
  if (parent.lastChild == targetElement) {
    parent.appendChild(newElement)
  } else {
    parent.insertBefore(newElement, targetElement.nextSibling)
  }
}
